### PR TITLE
adding gridfs support

### PR DESCRIPTION
**Pull request by diabolicallabs**
Signed-off-by: David Bush <dbush@redhat.com>

### CODE REVIEW

**Comment by vietj** [comment]
one of the test you added does not pass in the Travis CI

File: src/main/java/examples/MongoClientExamples.java [hunk 1]
```diff
@@ -15,13 +15,13 @@
...
+import io.vertx.core.AsyncResult;
 import io.vertx.core.Vertx;
+import io.vertx.core.file.AsyncFile;
 import io.vertx.core.json.JsonArray;
 import io.vertx.core.json.JsonObject;
 import io.vertx.docgen.Source;
-import io.vertx.ext.mongo.FindOptions;
-import io.vertx.ext.mongo.MongoClient;
-import io.vertx.ext.mongo.UpdateOptions;
+import io.vertx.ext.mongo.*;
```
Comment (karianna) [inline src/main/java/examples/MongoClientExamples.java]: prefer explicit imports

File: src/main/java/examples/MongoClientExamples.java [hunk 2]
```diff
@@ -414,4 +414,169 @@ public void example17_d1(MongoClient mongoClient) {
...
+  public void example26(MongoGridFsClient gridFsClient) {
+    gridFsClient.downloadFile("file.name", res -> {
+      if (res.succeeded()) {
+        Long fileLength = res.result();
+        //The length of the file stored in fileName
+      } else {
+        res.cause().printStackTrace();
+      }
+    });
+  }
```
Comment (karianna) [inline src/main/java/examples/MongoClientExamples.java]: whitespace between methods

File: src/main/java/examples/MongoClientExamples.java [hunk 3]
```diff
@@ -414,4 +414,169 @@ public void example17_d1(MongoClient mongoClient) {
...
+    gridFsClient.downloadFileByID(id, filename, res -> {
+      if (res.succeeded()) {
+        Long fileLength = res.result();
+        //The length of the file stored in fileName
+      } else {
+        res.cause().printStackTrace();
+      }
+    });
+  }
+  public void example28(MongoGridFsClient gridFsClient) {
```
Comment (karianna) [inline src/main/java/examples/MongoClientExamples.java]: whitespace between methods

File: src/main/java/io/vertx/ext/mongo/DownloadOptions.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package io.vertx.ext.mongo;
```
Comment (karianna) [inline src/main/java/io/vertx/ext/mongo/DownloadOptions.java]: I wonder if DownloadOptions is too generic and this needs to be named more specifically, e.g. That it's retrieval via GridFS

File: src/main/java/io/vertx/ext/mongo/UploadOptions.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
+package io.vertx.ext.mongo;
+
+import io.vertx.codegen.annotations.DataObject;
+import io.vertx.core.json.JsonObject;
+
+@DataObject
+public class UploadOptions {
```
Comment (karianna) [inline src/main/java/io/vertx/ext/mongo/UploadOptions.java]: See comment about DownlodOptions naming

File: src/main/java/io/vertx/ext/mongo/impl/GridFSInputStreamImpl.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    buffer = new CircularByteBuffer(DEFAULT_BUFFER_SIZE);
+    writeQueueMaxSize = buffer.capacity();
+  }
+
+  public GridFSInputStreamImpl(final int queueSize) {
+    buffer = new CircularByteBuffer(queueSize);
+    writeQueueMaxSize = queueSize;
+  }
+
+  public void read(ByteBuffer b, SingleResultCallback<Integer> c) {
```
Comment (karianna) [inline src/main/java/io/vertx/ext/mongo/impl/GridFSInputStreamImpl.java]: What is a 'b' what is a 'c'? :-)

Repeated below as well

File: src/main/java/io/vertx/ext/mongo/impl/GridFSInputStreamImpl.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    synchronized (buffer) {
```
Comment (karianna) [inline src/main/java/io/vertx/ext/mongo/impl/GridFSInputStreamImpl.java]: @johnoliver Can you think about the saneness of the concurrency here - it's fine by me but I'm o longer a practioner

File: src/main/java/io/vertx/ext/mongo/impl/GridFSInputStreamImpl.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+      }
+      this.closed = true;
+    }
+  }
+
+  public WriteStream<Buffer> exceptionHandler(Handler<Throwable> handler) {
+    return this;
+  }
+
+  public GridFSInputStream setWriteQueueMaxSize(int i) {
```
Comment (karianna) [inline src/main/java/io/vertx/ext/mongo/impl/GridFSInputStreamImpl.java]: i --> size

File: src/main/java/io/vertx/ext/mongo/impl/MongoGridFsClientImpl.java [hunk 9]
```diff
@@ -0,0 +1,316 @@
+package io.vertx.ext.mongo.impl;
+
+import com.mongodb.async.client.gridfs.GridFSBucket;
+import com.mongodb.client.gridfs.model.GridFSDownloadOptions;
+import com.mongodb.client.gridfs.model.GridFSUploadOptions;
+import io.vertx.core.*;
```
Comment (karianna) [inline src/main/java/io/vertx/ext/mongo/impl/MongoGridFsClientImpl.java]: prefer explicit imports

File: src/main/java/io/vertx/ext/mongo/util/CircularByteBuffer.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class CircularByteBuffer {
+
+  private byte[] array;
+  private int sizeMask;
+  private final AtomicInteger writeSequence = new AtomicInteger(0);
+  private final AtomicInteger readSequence = new AtomicInteger(0);
+
+  public CircularByteBuffer(int size) {
```
Comment (karianna) [inline src/main/java/io/vertx/ext/mongo/util/CircularByteBuffer.java]: Is there a 3rd party circular byte buffer in Guava or some other dependency that vertx already has?  Nothing wrong with this impl but prefer not to have data structures inside the driver which may have battle tested variants elsewhere.

Comment (diabolicallabs) [reply src/main/java/io/vertx/ext/mongo/util/CircularByteBuffer.java]: I'll give CircularFifoQueue from Apache Commons a try. If I'm not mistaken, Julien doesn't care much for dependencies on Guava.

File: src/main/java/examples/MongoClientExamples.java (hunk 1 above)
Comment (diabolicallabs) [reply src/main/java/examples/MongoClientExamples.java]: fixed

File: src/main/java/examples/MongoClientExamples.java (hunk 2 above)
Comment (diabolicallabs) [reply src/main/java/examples/MongoClientExamples.java]: fixed

File: src/main/java/examples/MongoClientExamples.java (hunk 3 above)
Comment (diabolicallabs) [reply src/main/java/examples/MongoClientExamples.java]: fixed

File: src/main/java/io/vertx/ext/mongo/DownloadOptions.java (hunk 4 above)
Comment (diabolicallabs) [reply src/main/java/io/vertx/ext/mongo/DownloadOptions.java]: Changed to GridFsDownloadOptions

File: src/main/java/io/vertx/ext/mongo/UploadOptions.java (hunk 5 above)
Comment (diabolicallabs) [reply src/main/java/io/vertx/ext/mongo/UploadOptions.java]: Changed to GridFsUploadOptions

File: src/main/java/io/vertx/ext/mongo/impl/GridFSInputStreamImpl.java (hunk 6 above)
Comment (diabolicallabs) [reply src/main/java/io/vertx/ext/mongo/impl/GridFSInputStreamImpl.java]: fixed

File: src/main/java/io/vertx/ext/mongo/impl/GridFSInputStreamImpl.java (hunk 8 above)
Comment (diabolicallabs) [reply src/main/java/io/vertx/ext/mongo/impl/GridFSInputStreamImpl.java]: fixed

File: src/main/java/io/vertx/ext/mongo/impl/MongoGridFsClientImpl.java (hunk 9 above)
Comment (diabolicallabs) [reply src/main/java/io/vertx/ext/mongo/impl/MongoGridFsClientImpl.java]: fixed

File: src/main/java/io/vertx/ext/mongo/util/CircularByteBuffer.java (hunk 10 above)
Comment (diabolicallabs) [reply src/main/java/io/vertx/ext/mongo/util/CircularByteBuffer.java]: Changed to Apache CircularFifoQueue

**Comment by karianna** [comment]
All good by me - over to @johnoliver

**Comment by kostya05983** [comment]
Can we finish with it? @johnoliver , this is cool feature for vertx-mongo-client

**Comment by karianna** [comment]
merge conflicts

**Comment by diabolicallabs** [comment]
Doing the merge sooner would have been better. I've moved on to Quarkus but I'll see what I can do in the next few days.

**Comment by kostya05983** [comment]
@diabolicallabs , maybe you can give me a write access to your fork, and I will be try to resolve conflicts or I suggest a PR to your fork to resolve conflicts

**Comment by diabolicallabs** [comment]
Thanks Konstantin. I like to finish what I start. But, if I don't get to it by Friday, I'll give you full access.

**Comment by kostya05983** [comment]
I made a rebase and migrate tests to promises. I think it's ready for review.  @karianna.

**Comment by vietj** [comment]
thanks @diabolicallabs and @kostya05983  for the contributions, this is much appreciated

**Pull request merged**